IPython/utils/traitlets.py

```diff
@@ -1069,7 +1069,9 @@ def instance_init(self, obj):
     def validate(self, obj, value):
         for trait_type in self.trait_types:
             try:
-                return trait_type._validate(obj, value)
+                v = trait_type._validate(obj, value)
+                self._metadata = v._metadata
```

Comment:
it sets metadata from *each* of the union traits, so Union(A, B) results in different metadata than Union(B, A).